File: backend/db/crud.py
from sqlalchemy.orm import Session
import uuid
from backend.db.models import User, Chain, Ring, Key, Charm
from backend.db import constants

# Basic CRUD operations for User, Chain, Ring, Key, and Charm


# User CRUD operations
def create_user(db: Session, email: str, display_name: str) -> User | None:
    user = User(email=email, display_name=display_name)
    db.add(user)
    db.commit()
    db.refresh(user)
    return user


def get_user(db: Session, user_id: uuid.UUID) -> User | None:
    return db.query(User).filter(User.id == user_id).first()


def get_user_by_email(db: Session, email: str) -> User | None:
    return db.query(User).filter(User.email == email).first()


def get_all_users(db: Session) -> list[User]:
    return db.query(User).all()


def update_user(
    db: Session,
    user_id: uuid.UUID,
    email: str | None = None,
    display_name: str | None = None,
) -> User | None:
    user = get_user(db, user_id)
    if user:
        if email:
            user.email = email
        if display_name:
            user.display_name = display_name
        db.commit()
        db.refresh(user)
    return user


def delete_user(db: Session, user_id: uuid.UUID) -> bool:
    user = get_user(db, user_id)
    if user:
        db.delete(user)
        db.commit()
        return True
    return False


# Chain CRUD operations
def create_chain(
    db: Session,
    user_id: uuid.UUID,
    name: str = "chain",
    root_id: uuid.UUID | None = None,
) -> Chain | None:
    chain = Chain(user_id=user_id, name=name, root_id=root_id)
    db.add(chain)
    db.commit()
    db.refresh(chain)
    return chain


def get_chain(db: Session, chain_id: uuid.UUID) -> Chain | None:
    return db.query(Chain).filter(Chain.id == chain_id).first()


def get_chains(db: Session, user_id: uuid.UUID | None = None) -> list[Chain]:
    """Return chains optionally filtered by user_id."""
    query = db.query(Chain)
    if user_id:
        query = query.filter(Chain.user_id == user_id)
    return query.all()


def update_chain(
    db: Session, chain_id: uuid.UUID, name: str | None = None, root_id: uuid.UUID | None = None
) -> Chain | None:
    chain = get_chain(db, chain_id)
    if chain:
        if name:
            chain.name = name
        if root_id:
            chain.root_id = root_id
        db.commit()
        db.refresh(chain)
    return chain


def delete_chain(db: Session, chain_id: uuid.UUID) -> bool:
    chain = get_chain(db, chain_id)
    if chain:
        db.delete(chain)
        db.commit()
        return True
    return False


# Chain CRUD operations
def create_ring(
    db: Session,
    chain_id: uuid.UUID,
    name: str = "ring",
    color: str = constants.RING_COLORS[0],
    parent_id: uuid.UUID | None = None,
) -> Ring | None:
    ring = Ring(chain_id=chain_id, name=name, color=color, parent_id=parent_id)
    db.add(ring)
    db.commit()
    db.refresh(ring)
    return ring


def get_ring(db: Session, ring_id: uuid.UUID) -> Ring | None:
    return db.query(Ring).filter(Ring.id == ring_id).first()


def get_rings(db: Session, chain_id: uuid.UUID | None = None, parent_id: uuid.UUID | None = None) -> list[Ring]:
    """Return rings optionally filtered by chain_id and/or parent_id."""
    query = db.query(Ring)
    if chain_id:
        query = query.filter(Ring.chain_id == chain_id)
    if parent_id is not None:
        query = query.filter(Ring.parent_id == parent_id)
    return query.all()


def update_ring(
    db: Session,
    ring_id: uuid.UUID,
    name: str | None = None,
    color: str | None = None,
    parent_id: uuid.UUID | None = None,
) -> Ring | None:
    ring = get_ring(db, ring_id)
    if ring:
        if name:
            ring.name = name
        if color:
            ring.color = color
        if parent_id is not None:
            ring.parent_id = parent_id
        db.commit()
        db.refresh(ring)
    return ring


def delete_ring(db: Session, ring_id: uuid.UUID) -> bool:
    ring = get_ring(db, ring_id)
    if ring:
        db.delete(ring)
        db.commit()
        return True
    return False


# Key CRUD operations
def create_key(
    db: Session,
    chain_id: uuid.UUID,
    name: str = "key",
    color: str = constants.KEY_COLORS[0],
    parent_id: uuid.UUID | None = None,
) -> Key | None:
    key = Key(chain_id=chain_id, name=name, color=color, parent_id=parent_id)
    db.add(key)
    db.commit()
    db.refresh(key)
    return key


def get_key(db: Session, key_id: uuid.UUID) -> Key | None:
    return db.query(Key).filter(Key.id == key_id).first()


def get_keys(db: Session, chain_id: uuid.UUID | None = None, parent_id: uuid.UUID | None = None) -> list[Key]:
    """Return keys optionally filtered by chain_id and/or parent_id."""
    query = db.query(Key)
    if chain_id:
        query = query.filter(Key.chain_id == chain_id)
    if parent_id is not None:
        query = query.filter(Key.parent_id == parent_id)
    return query.all()


def update_key(
    db: Session,
    key_id: uuid.UUID,
    name: str | None = None,
    color: str | None = None,
    parent_id: uuid.UUID | None = None,
) -> Key | None:
    key = get_key(db, key_id)
    if key:
        if name:
            key.name = name
        if color:
            key.color = color
        if parent_id is not None:
            key.parent_id = parent_id
        db.commit()
        db.refresh(key)
    return key


def delete_key(db: Session, key_id: uuid.UUID) -> bool:
    key = get_key(db, key_id)
    if key:
        db.delete(key)
        db.commit()
        return True
    return False


# Charm CRUD operations
def create_charm(
    db: Session,
    chain_id: uuid.UUID,
    name: str = "charm",
    type: str = constants.CHARM_TYPES[0],
    parent_id: uuid.UUID | None = None,
) -> Charm | None:
    charm = Charm(chain_id=chain_id, name=name, type=type, parent_id=parent_id)
    db.add(charm)
    db.commit()
    db.refresh(charm)
    return charm


def get_charm(db: Session, charm_id: uuid.UUID) -> Charm | None:
    return db.query(Charm).filter(Charm.id == charm_id).first()


def get_charms(db: Session, chain_id: uuid.UUID | None = None, parent_id: uuid.UUID | None = None) -> list[Charm]:
    """Return charms optionally filtered by chain_id and/or parent_id."""
    query = db.query(Charm)
    if chain_id:
        query = query.filter(Charm.chain_id == chain_id)
    if parent_id is not None:
        query = query.filter(Charm.parent_id == parent_id)
    return query.all()


def update_charm(
    db: Session,
    charm_id: uuid.UUID,
    name: str | None = None,
    type: str | None = None,
    parent_id: uuid.UUID | None = None,
) -> Charm | None:
    charm = get_charm(db, charm_id)
    if charm:
        if name:
            charm.name = name
        if type:
            charm.type = type
        if parent_id is not None:
            charm.parent_id = parent_id
        db.commit()
        db.refresh(charm)
    return charm


def delete_charm(db: Session, charm_id: uuid.UUID) -> bool:
    charm = get_charm(db, charm_id)
    if charm:
        db.delete(charm)
        db.commit()
        return True
    return False

File: frontend/src/lib/functions/crud.ts
// Types
export interface UserCreate {
  email: string;
  display_name: string;
}

export interface UserRead {
  id: string;
  email: string;
  display_name: string;
}

export interface ChainCreate {
  user_id: string;
  name: string;
}

export interface ChainRead {
  id: string;
  user_id: string;
  name: string;
  root_id: string | null;
}

export interface RingCreate {
  chain_id: string;
  name: string;
  color: string;
  parent_id?: string | null;
}

export interface RingRead {
  id: string;
  chain_id: string;
  name: string;
  color: string;
  parent_id: string | null;
}

export interface KeyCreate {
  chain_id: string;
  name: string;
  color: string;
  parent_id?: string | null;
}

export interface KeyRead {
  id: string;
  chain_id: string;
  name: string;
  color: string;
  parent_id: string | null;
}

export interface CharmCreate {
  chain_id: string;
  name: string;
  type: string;
  parent_id?: string | null;
}

export interface CharmRead {
  id: string;
  chain_id: string;
  name: string;
  type: string;
  parent_id: string | null;
}

// API Client
class KeychainAPI {
  private baseUrl: string;

  constructor(baseUrl: string = "http://localhost:8000") {
    this.baseUrl = baseUrl;
  }

  private async request<T>(
    endpoint: string,
    options?: RequestInit
  ): Promise<T> {
    const res = await fetch(`${this.baseUrl}${endpoint}`, {
      ...options,
      headers: {
        "Content-Type": "application/json",
        ...options?.headers,
      },
    });

    if (!res.ok) {
      const error = await res.json().catch(() => ({ detail: res.statusText }));
      throw new Error(error.detail || `HTTP ${res.status}`);
    }

    // Some endpoints (DELETE) return 204 No Content. Avoid calling res.json()
    // when there's no body.
    if (res.status === 204) {
      return undefined as unknown as T;
    }

    return res.json();
  }

  // Status
  async getStatus(): Promise<{ service: string; status: string }> {
    return this.request("/");
  }

  // Users
  async createUser(user: UserCreate): Promise<UserRead> {
    return this.request("/users", {
      method: "POST",
      body: JSON.stringify(user),
    });
  }

  async getUser(userId: string): Promise<UserRead> {
    return this.request(`/users/${userId}`);
  }

  async getUserByEmail(email: string): Promise<UserRead> {
    const users = await this.request<UserRead[]>(
      `/users?email=${encodeURIComponent(email)}`
    );
    if (users.length === 0) {
      throw new Error("User not found");
    }
    return users[0];
  }

  async listUsers(): Promise<UserRead[]> {
    return this.request("/users");
  }

  // Chains
  async createChain(chain: ChainCreate): Promise<ChainRead> {
    return this.request("/chains", {
      method: "POST",
      body: JSON.stringify(chain),
    });
  }

  async getChain(chainId: string): Promise<ChainRead> {
    return this.request(`/chains/${chainId}`);
  }

  async listChains(userId: string): Promise<ChainRead[]> {
    return this.request(`/chains?user_id=${userId}`);
  }

  // Rings
  async createRing(ring: RingCreate): Promise<RingRead> {
    return this.request("/rings", {
      method: "POST",
      body: JSON.stringify(ring),
    });
  }

  async getRing(ringId: string): Promise<RingRead> {
    return this.request(`/rings/${ringId}`);
  }

  async listRings(params?: {
    chain_id?: string;
    parent_id?: string;
  }): Promise<RingRead[]> {
    const query = new URLSearchParams();
    if (params?.chain_id) query.set("chain_id", params.chain_id);
    if (params?.parent_id) query.set("parent_id", params.parent_id);
    const queryStr = query.toString();
    return this.request(`/rings${queryStr ? `?${queryStr}` : ""}`);
  }

  // Keys
  async createKey(key: KeyCreate): Promise<KeyRead> {
    return this.request("/keys", {
      method: "POST",
      body: JSON.stringify(key),
    });
  }

  async getKey(keyId: string): Promise<KeyRead> {
    return this.request(`/keys/${keyId}`);
  }

  async listKeys(params?: {
    chain_id?: string;
    parent_id?: string;
  }): Promise<KeyRead[]> {
    const query = new URLSearchParams();
    if (params?.chain_id) query.set("chain_id", params.chain_id);
    if (params?.parent_id) query.set("parent_id", params.parent_id);
    const queryStr = query.toString();
    return this.request(`/keys${queryStr ? `?${queryStr}` : ""}`);
  }

  // Charms
  async createCharm(charm: CharmCreate): Promise<CharmRead> {
    return this.request("/charms", {
      method: "POST",
      body: JSON.stringify(charm),
    });
  }

  async getCharm(charmId: string): Promise<CharmRead> {
    return this.request(`/charms/${charmId}`);
  }

  async listCharms(params?: {
    chain_id?: string;
    parent_id?: string;
  }): Promise<CharmRead[]> {
    const query = new URLSearchParams();
    if (params?.chain_id) query.set("chain_id", params.chain_id);
    if (params?.parent_id) query.set("parent_id", params.parent_id);
    const queryStr = query.toString();
    return this.request(`/charms${queryStr ? `?${queryStr}` : ""}`);
  }

  // DELETE helpers
  async deleteRing(ringId: string): Promise<void> {
    return this.request<void>(`/rings/${ringId}`, { method: "DELETE" });
  }

  async deleteKey(keyId: string): Promise<void> {
    return this.request<void>(`/keys/${keyId}`, { method: "DELETE" });
  }

  async deleteCharm(charmId: string): Promise<void> {
    return this.request<void>(`/charms/${charmId}`, { method: "DELETE" });
  }
}

// Export the class
export { KeychainAPI };

// Usage example:
// const api = new KeychainAPI('http://localhost:8000');
// const user = await api.createUser({ email: 'test@example.com', display_name: 'Test' });
// const chain = await api.createChain({ user_id: user.id, name: 'My Chain' });
// const rings = await api.listRings({ chain_id: chain.id });
